### DISCUSSION TITLE
[question] How can I compare function signatures in the `typing.Callable` syntax to `inspect.signature` syntax>

### DISCUSSION THREAD

**Discussion by sg-s** [python/typing#991] [Q&A]
My ultimate goal is to enforce some sort of type for functions that are passed to arguments to other functions.

I have a function that looks like:

```python
def foo(bar: Callable[[int], int]):
```

I want to implement functionality so that it only accepts functions that look like:

```python
def bar(a: int)->int:
```

If I `inspect` bar, I get this:

```python
inspect.signature(bar)
<Signature (a: int) -> int>
```

How do I work with this and make sure it's the same as the function signature as defined in the `Callable` syntax?

**Comment by JelleZijlstra** [comment]
The native typing libraries aren't geared towards supporting this kind of runtime introspection. However, libraries that perform runtime introspection of types should support this. Examples include beartype and pydantic and the type checker I wrote, pyanalyze.

Here's how you could check this with pyanalyze:

```
from pyanalyze.annotations import type_from_runtime
from pyanalyze.test_value import CTX
from pyanalyze.value import KnownValue
from typing import Callable

cv = type_from_runtime(Callable[[int], int])


def bar(a: int) -> int:
    pass


def bad(a: str) -> int:
    pass

# ok
print(cv.can_assign(KnownValue(bar), CTX)) #  {} (indicating compatibility)

# error
#  type of positional-only parameter '__arg0' is incompatible
#    Cannot assign int to str
print(cv.can_assign(KnownValue(bad), CTX))
```

**Reply by sg-s** [reply]
thank you -- does pyanalyze have a wrapper that i can use to enforce type annotations?

**Reply by hauntsaninja** [reply]
https://github.com/agronholm/typeguard is also a pretty popular option here

**Comment by JelleZijlstra** [comment]
No, it’s not designed for runtime typechecking. The other tools I mentioned
do, though.
